docs/web3.contract.rst

```diff
@@ -296,54 +314,156 @@ Each Contract Factory exposes the following methods.
+        >>> filter_builder = contract.events.Transfer.build_filter()
+        >>> filter_builder.from_block = "latest"
+        >>> filter_builder.from_block = 0
+        Traceback (most recent call last):
+        web3.exceptions.Web3ValueError: from_block is already set to 'latest'. Resetting filter parameters is not permitted
 
 .. py:classmethod:: Contract.encode_abi(abi_element_identifier, args=None, kwargs=None, data=None)
 
     Encodes the arguments using the Ethereum ABI for the contract function that
     matches the given ``abi_element_identifier`` and arguments ``args``. The ``data`` parameter
     defaults to the function selector.
 
-    .. code-block:: python
+    .. doctest:: contractmethods
+
+      >>> contract.encode_abi("approve", args=[w3.eth.accounts[0], 10])
+      '0x095ea7b30000000000000000000000007e5f4552091a69125d5dfcb7b8c2659029395bdf000000000000000000000000000000000000000000000000000000000000000a'
+
+.. py:classmethod:: Contract.all_events()
+
+    Returns a list of all the events present in a Contract where every event is
+    an instance of :py:class:`ContractEvent`.
+
+    .. doctest:: contractmethods
+
+        >>> contract.all_events()
+        [<Event Transfer(address,address,uint256)>, <Event Approval(address,address,uint256)>]
+
+
+.. py:classmethod:: Contract.get_event_by_signature(signature)
+
+    Searches for a distinct event with matching signature. Returns an instance of
+    :py:class:`ContractEvent` upon finding a match. Raises ``Web3ValueError`` if no
+    match is found.
+
+    .. doctest:: contractmethods
+
+        >>> contract.get_event_by_signature('Transfer(address,address,uint256)')
+        <Event Transfer(address,address,uint256)>
+
+
+.. py:classmethod:: Contract.find_events_by_name(name)
 
-      >>> contract.encode_abi("register", args=["rainbows", 10])
-      "0xea87152b0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000087261696e626f7773000000000000000000000000000000000000000000000000"
+    Searches for all events matching the provided name. Returns a list of matching
+    events where every event is an instance of :py:class:`ContractEvent`. Returns an
+    empty list when no match is found.
 
+    .. doctest:: contractmethods
+
+        >>> contract.find_events_by_name('Transfer')
+        [<Event Transfer(address,address,uint256)>]
+
+
+.. py:classmethod:: Contract.get_event_by_name(name)
+
+    Searches for a distinct event matching the name. Returns an instance of
+    :py:class:`ContractEvent` upon finding a match. Raises ``Web3ValueError`` if no
+    match is found or if multiple matches are found.
+
+    .. doctest:: contractmethods
+
+        >>> contract.get_event_by_name('Approval')
+        <Event Approval(address,address,uint256)>
+
+.. py:classmethod:: Contract.find_events_by_selector(selector)
```

Comment:
Does this work with the 4byte selector too or just the whole thing?

Reply:
I'm not certain I've come across events that are represented by a 4byte selector, so this is implemented to match the whole event signature encoding.

Reply:
Oh, right. This is events. I think we can just leave it as-is. If we get a request to change it later, we can.